server/src/linter.ts

```diff
@@ -0,0 +1,198 @@
+import { spawn } from 'child_process'
+import * as LSP from 'vscode-languageserver'
+
+function formatMessage(comment: ShellcheckComment): string {
+  return (comment.code ? `SC${comment.code}: ` : '') + comment.message
+}
+
+type LinterOptions = {
+  executablePath?: string | null
+  cwd?: string
+}
+
+export default class Linter {
+  private executablePath: string | null
+  private cwd: string
+  _canLint: boolean
+
+  constructor(opts: LinterOptions = {}) {
```

Comment:
Why not force options to be there?

```suggestion
  constructor(opts: LinterOptions) {
    this.executablePath = opts.executablePath
```